File: python/src/azoth/eos/reference/matcop_prumr_alpha.py
"""``eos.matcop_prumr_alpha`` - the Mathias-Copeman alpha with the UMR-PR fallback.

Spec: ``specs/calcs/eos/matcop_prumr_alpha.toml``, which carries the provenance and why
the three coefficients are the caller's rather than derived.
"""

from __future__ import annotations

import math

from azoth._registry_gen import spec as _spec_for
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import MatcopPrumrAlphaResult
from azoth.core.warnings import Warning

CALC_ID = "eos.matcop_prumr_alpha"


def matcop_prumr_alpha(
    omega: float, mc1: float, mc2: float, mc3: float, Tr: float
) -> MatcopPrumrAlphaResult:
    """The Mathias-Copeman alpha function for a pure component, with the standard
    Peng-Robinson alpha when the first coefficient is essentially zero.

    Args:
        omega: the acentric factor, used only for the fallback.
        mc1: the first Mathias-Copeman coefficient.
        mc2: the second Mathias-Copeman coefficient.
        mc3: the third Mathias-Copeman coefficient.
        Tr: the reduced temperature.

    Returns:
        The temperature-dependent alpha function.

    Raises:
        OutOfRangeError: if ``Tr`` is not positive.

    Example:
        >>> matcop_prumr_alpha(0.1, 0.5, 0.2, -0.1, 0.7).alpha
        1.180634768967036
    """
    spec = _spec_for(CALC_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    values = {"omega": omega, "mc1": mc1, "mc2": mc2, "mc3": mc3, "Tr": Tr}
    apply_checks(checks.on_input, values.get, warnings)

    if mc1 < 1e-20:
        m = (
            0.384401
            + 1.52276 * omega
            - 0.213808 * omega * omega
            + 0.034616 * omega * omega * omega
            - 0.001976 * omega * omega * omega * omega
        )
        t = 1.0 + m * (1.0 - math.sqrt(Tr))
        alpha = t * t
    else:
        root = 1.0 - math.sqrt(Tr)
        t = 1.0 + mc1 * root + mc2 * root * root + mc3 * root * root * root
        alpha = t * t

    apply_checks(checks.derived, lambda name: alpha if name == "alpha" else None, warnings)

    return MatcopPrumrAlphaResult(alpha=alpha, warnings=tuple(warnings))

File: crates/azoth-eos/src/matcop_prumr_alpha.rs
//! `eos.matcop_prumr_alpha` - the Mathias-Copeman alpha function with the UMR-PR
//! fallback.
//!
//! ```text
//! alpha = (1 + mc1*(1 - sqrt(Tr)) + mc2*(1 - sqrt(Tr))**2 + mc3*(1 - sqrt(Tr))**3)**2
//!         falling back to (1 + m*(1 - sqrt(Tr)))**2 when mc1 < 1e-20
//! m     = 0.384401 + 1.52276*omega - 0.213808*omega**2 + 0.034616*omega**3
//!         - 0.001976*omega**4
//! ```
//!
//! Spec: `specs/calcs/eos/matcop_prumr_alpha.toml`, which carries the provenance and why
//! the three coefficients are the caller's rather than derived.

use azoth_core::{Result, apply_checks};

use crate::results::MatcopPrumrAlphaResult;
use crate::spec_gen;

/// The Mathias-Copeman alpha function for a pure component, with the standard
/// Peng-Robinson alpha when the first coefficient is essentially zero.
///
/// `mc1`, `mc2` and `mc3` are the fitted Mathias-Copeman coefficients; `omega` the
/// acentric factor, used only for the fallback; `Tr` the reduced temperature.
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `Tr <= 0` - the formula takes a square
///   root of `Tr`.
///
/// # Example
/// ```
/// use azoth_eos::matcop_prumr_alpha;
///
/// let r = matcop_prumr_alpha(0.1, 0.5, 0.2, -0.1, 0.7)?;
/// assert!((r.alpha - 1.180634768967036).abs() < 1e-15);
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
#[allow(non_snake_case)] // `Tr` is the symbol in the published equation
pub fn matcop_prumr_alpha(
    omega: f64,
    mc1: f64,
    mc2: f64,
    mc3: f64,
    Tr: f64,
) -> Result<MatcopPrumrAlphaResult> {
    let spec = &spec_gen::MATCOP_PRUMR_ALPHA_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "omega" => Some(omega),
            "mc1" => Some(mc1),
            "mc2" => Some(mc2),
            "mc3" => Some(mc3),
            "Tr" => Some(Tr),
            _ => None,
        },
        &mut warnings,
    )?;

    let alpha = if mc1 < 1e-20 {
        let m = 0.384_401 + 1.522_76 * omega - 0.213_808 * omega * omega
            + 0.034_616 * omega * omega * omega
            - 0.001_976 * omega * omega * omega * omega;
        let t = 1.0 + m * (1.0 - Tr.sqrt());
        t * t
    } else {
        let root = 1.0 - Tr.sqrt();
        let t = 1.0 + mc1 * root + mc2 * root * root + mc3 * root * root * root;
        t * t
    };

    apply_checks(
        spec.derived_checks(),
        |name| (name == "alpha").then_some(alpha),
        &mut warnings,
    )?;

    Ok(MatcopPrumrAlphaResult { alpha, warnings })
}
